### PR TITLE
Bug/402

### PR DESCRIPTION
**Pull request by OlegGordiichuk**
Rewrite logic for showing of the passcode. I take a look at apps that are in GitHub and use their approach. Add unit tests. Test on iPhone (10.3) and iPad (11).

### CODE REVIEW

File: Trust/Lock/Coordinators/LockEnterPasscodeCoordinator.swift [hunk 1]
```diff
@@ -13,28 +13,27 @@ class LockEnterPasscodeCoordinator: Coordinator {
...
                 self?.stop()
             }
+            self?.protectionWasShown = bioUnlock
         }
     }
     func start() {
         guard lock.isPasscodeSet() else {
-             //Due to the lazy init and navigation flow we should call viewWillAppear manually.
-            lockEnterPasscodeViewController.viewWillAppear(true)
             return
```
Comment (michaelScoff) [inline Trust/Lock/Coordinators/LockEnterPasscodeCoordinator.swift]: should we be calling stop method if passcode not set?

File: Trust/Protection/Coordinators/ProtectionCoordinator.swift [hunk 2]
```diff
@@ -12,24 +12,25 @@ class ProtectionCoordinator: Coordinator {
     }()
     let protectionWindow: UIWindow
     init() {
+        let protection = UIWindow()
```
Comment (michaelScoff) [inline Trust/Protection/Coordinators/ProtectionCoordinator.swift]: I think you can move initializer on top `let protectionWindow = UIWindow()`

File: Trust/Protection/Coordinators/SplashCoordinator.swift [hunk 3]
```diff
@@ -4,24 +4,25 @@ import UIKit
...
+            splashView.leadingAnchor.constraint(equalTo: window.leadingAnchor),
+            splashView.bottomAnchor.constraint(equalTo: window.bottomAnchor),
+            splashView.trailingAnchor.constraint(equalTo: window.trailingAnchor),
+        ])
     }
-    func dismiss() {
-        splashViewIsActive = false
-        window.isHidden = true
+    func stop() {
+        splashView.removeFromSuperview()
```
Comment (michaelScoff) [inline Trust/Protection/Coordinators/SplashCoordinator.swift]: Do you need to hide window as well?

**Comment by michaelScoff** [comment]
Can you add tests for `LockEnterPasscodeCoordinator` where user disable or enable during using the app. Just want to make sure we cover all the cases

File: Trust/Protection/Coordinators/SplashCoordinator.swift (hunk 3 above)
Comment (OlegGordiichuk) [reply Trust/Protection/Coordinators/SplashCoordinator.swift]: No splash screen we add like subview.

File: TrustTests/Coordinators/SplashCoordinatorTests.swift [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        coordinator.start()
+        XCTAssertFalse(window.isHidden)
+        XCTAssertNotNil(window.subviews.first)
+    }
+    func testStop() {
+        let window = UIWindow()
+        let coordinator = SplashCoordinator(window: window)
+        coordinator.start()
+        coordinator.stop()
+        XCTAssertNil(window.subviews.first)
```
Comment (michaelScoff) [inline TrustTests/Coordinators/SplashCoordinatorTests.swift]: Can you check if it's hidden as well on stop? Of if it doesn't have subviews than that's same thing?

File: TrustTests/Coordinators/SplashCoordinatorTests.swift [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+@testable import Trust
+
+
+class SplashCoordinatorTests: XCTestCase {
+    func testStart() {
+        let window = UIWindow()
+        let coordinator = SplashCoordinator(window: window)
+        coordinator.start()
+        XCTAssertFalse(window.isHidden)
+        XCTAssertNotNil(window.subviews.first)
```
Comment (michaelScoff) [inline TrustTests/Coordinators/SplashCoordinatorTests.swift]: One thing to check levels as well?

**Pull request merged**